From 106c3ea838dd4c38e8cdeeaa032abc70742e57a7 Mon Sep 17 00:00:00 2001
From: "Wu, Hao" <hao.wu@intel.com>
Date: Fri, 17 Oct 2014 13:38:38 +0800
Subject: [PATCH 010/441] phy-intel-cht: fixing notifier initialization issue

register notifier later than usb_add_phy function to avoid
issue that it is re-init again in usb_add_phy function.
and also change otg_id debugging interface to send otg events.

Change-Id: Idf5e9b4ea8137a30d3653dce5b09a81190b99050
Tracked-On: https://jira01.devtools.intel.com/browse/GMINL-2558
Signed-off-by: Wu, Hao <hao.wu@intel.com>
---
 drivers/usb/phy/phy-intel-cht.c | 15 +++++++--------
 1 file changed, 7 insertions(+), 8 deletions(-)

--- a/drivers/usb/phy/phy-intel-cht.c
+++ b/drivers/usb/phy/phy-intel-cht.c
@@ -365,15 +365,15 @@ static ssize_t store_otg_id(struct devic
 	case 'a':
 	case 'A':
 		dev_info(cht_otg_dev->phy.dev, "ID = 0\n");
-		fsm->id = 0;
-		otg_statemachine(fsm);
+		atomic_notifier_call_chain(&cht_otg_dev->phy.notifier,
+			USB_EVENT_ID, NULL);
 		return count;
 	case '1':
 	case 'b':
 	case 'B':
 		dev_info(cht_otg_dev->phy.dev, "ID = 1\n");
-		fsm->id = 1;
-		otg_statemachine(fsm);
+		atomic_notifier_call_chain(&cht_otg_dev->phy.notifier,
+			USB_EVENT_NONE, NULL);
 		return count;
 	default:
 		return -EINVAL;
@@ -416,10 +416,6 @@ static int cht_otg_probe(struct platform
 	cht_otg->phy.otg->start_hnp = NULL;
 	cht_otg->phy.otg->start_srp = NULL;
 
-	ATOMIC_INIT_NOTIFIER_HEAD(&cht_otg->phy.notifier);
-
-	cht_otg->nb.notifier_call = cht_otg_handle_notification;
-	usb_register_notifier(&cht_otg->phy, &cht_otg->nb);
 
 	INIT_WORK(&cht_otg->fsm_work, cht_otg_fsm_work);
 
@@ -431,6 +427,9 @@ static int cht_otg_probe(struct platform
 		goto err1;
 	}
 
+	cht_otg_dev->nb.notifier_call = cht_otg_handle_notification;
+	usb_register_notifier(&cht_otg_dev->phy, &cht_otg_dev->nb);
+
 	/* init otg-fsm */
 	status = cht_otg_start(pdev);
 	if (status) {
